src/main/java/com/sekwah/advancedportals/bukkit/portals/Portal.java

```diff
@@ -384,17 +385,26 @@ public static boolean activate(Player player, AdvancedPortal portal) {
             return false;
         }
 
-        if (cooldown.get(player.getName()) != null) {
-            int diff = (int) ((System.currentTimeMillis() - cooldown.get(player.getName())) / 1000);
-            if (diff < cooldelay) {
-                int time = (cooldelay - diff);
-                player.sendMessage(ChatColor.RED + "Please wait " + ChatColor.YELLOW + time + ChatColor.RED + (time == 1 ? " second" : " seconds") + " until attempting to enter this portal again.");
-                failSound(player, portal);
-                throwPlayerBack(player);
-                return false;
+        HashMap<String, Long> cds = cooldown.get(player.getName());
+        if (cds != null) {
+            if (cds.get(portal.getName()) != null) {
+                long cd = cds.get(portal.getName());
+                int diff = (int) ((System.currentTimeMillis() - cd) / 1000);
+                if (diff < portal.getCooldownDelay()) {
+                    int time = (portal.getCooldownDelay() - diff);
+                    player.sendMessage(ChatColor.RED + "Please wait " + ChatColor.YELLOW + time + ChatColor.RED + (time == 1 ? " second" : " seconds") + " until attempting to enter this portal again.");
+                    failSound(player, portal);
+                    throwPlayerBack(player);
+                    return false;
+                }
             }
         }
-        cooldown.put(player.getName(), System.currentTimeMillis());
+        if(cds == null) {
+            cds = new HashMap<String, Long>();
+            player.sendMessage("cds is set");
```

Comment:
Please avoid sending debug statements to the player.

I assume you missed this, I usually just use system.out to avoid players seeing debug statements that slip through the cracks